File: src/leetcode/07-22-24/maximum-depth-of-binary-tree.ts
/*
    Link: https://leetcode.com/problems/maximum-depth-of-binary-tree/

    Time Took: 8 minutes 

    Time Complexity: O(N)

    Space Complexity: O(N)

    Solution Explained: You can do bfs or dfs for this. For bfs you would traverse the tree level by level.
        The number of levels if your answer. For a dfs solution you do a bubble up approach.
        At each node, the depth of the longest binary tree starting at that node, is 
        1 + Math.max(depth(left), depth(right))
*/

namespace MaximumDepthOfBinaryTree {
    class TreeNode {
        val: number
        left: TreeNode | null
        right: TreeNode | null
        constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
            this.val = (val===undefined ? 0 : val)
            this.left = (left===undefined ? null : left)
            this.right = (right===undefined ? null : right)
        }
   }
   
   function maxDepth(root: TreeNode | null): number {
       if (!root)
           return 0;
   
       const leftDepth: number = maxDepth(root.left);
       const rightDepth: number = maxDepth(root.right);
   
       return 1 + Math.max(leftDepth, rightDepth);
   }
}

File: src/leetcode/neetcode-150/tree/maximum-depth-of-binary-tree.py
'''
    Link: https://leetcode.com/problems/maximum-depth-of-binary-tree/

    Time Took: 11 Minutes.

    Time Complexity: O(N)

    Space Complexity: O(N)

    Date: 03/03/2025

    Problem Type: tree

    Solution Explained: start at root node. have helper func that has as a param the current depth. As you visit nodes you increment that param.
    So everytime you are at a node, you know it's depth, compare that depth to the res depth that lives on the class instance.

    Notes:
'''

'''
need to traverse the entire tree
count as we get deeper 

dfs or bfs?

s1:
    dfs
'''

class TreeNode:
    def __init__(self, left, right, val):
        self.left = left
        self.right = right
        self.val = val

from typing import Optional

class Solution:
    resDepth: int = 0
    
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0

        self.maxDepthHelper(root, 1)
        return self.resDepth
    
    def maxDepthHelper(self, root: Optional[TreeNode], currentDepth: int):
        if root == None:
            return

        self.resDepth = max(currentDepth, self.resDepth)

        self.maxDepthHelper(root.left, currentDepth + 1)
        self.maxDepthHelper(root.right, currentDepth + 1)
